### DISCUSSION TITLE
Confused about how to pass arguments to a scala (sbt) job

### DISCUSSION THREAD

**Discussion by jsatk** [mfussenegger/nvim-dap#272] [Q&A] [answered]
Firstly I have the following in my `~/.config/nvim/init.lua` (slightly truncated for readability).

```lua
metals_config.on_attach = function(client, bufnr)
  require("metals").setup_dap()
end

dap.configurations.scala = {
  {
    type = "scala",
    request = "launch",
    name = "Run",
    metals = {
      runType = "run"
    }
  },
  {
    type = "scala",
    request = "launch",
    name = "Test File",
    metals = {
      runType = "testFile"
    }
  },
  {
    type = "scala",
    request = "launch",
    name = "Test Target",
    metals = {
      runType = "testTarget"
    }
  },
}
```

When I run `lua require"dap".continue()<CR>` (which I have mapped to `<leader>dc` it prompts me for what configuration I want to use. I input 1 for `Run`. And then it brings up a list of jobs defined in my `build.sbt`. I choose one of those jobs and that's it. It says it's "Compiling root" and then nothing happens. The command I'm trying to run in particular requires some arguments.

What is the right way to run this sbt command with arguments and hit the breakpoint I've set?

Thank you in advance.

Note: I previously asked a very similar question here. The one thing I never got out of it was how to pass arguments to jobs though. https://github.com/mfussenegger/nvim-dap/issues/162

**Comment by ckipp01** [comment, marked as answer]
> And then it brings up a list of jobs defined in my build.sbt

Just to clarify, but jobs here do you mean main methods? If you've triggered a `Run` here then Metals will either automatically run the build target you're in if there is only one main method _or_ will give you a choice to choose with main method you'd like to run.

> I choose one of those jobs and that's it. It says it's "Compiling root" and then nothing happens. The command I'm trying to run in particular requires some arguments.

You may also want to toggle your repl when you're doing this if you're not already via `require"dap".repl.toggle()`.

As for the args that you want to pass into run, you'll actually include them in the `metals` key. Foe example your run configuration would look something like this:

```lua
 {
    type = "scala",
    request = "launch",
    name = "Run",
    metals = {
      runType = "run",
      args = { "firstArg", "secondArg", "thirdArg" }
    }
  }

**Reply by jsatk** [reply]
> Just to clarify, but jobs here do you mean main methods?

I do. Apologies for using the wrong terminology.

Ah. So when I am prompted and press 1 for `Run` I see this (job names edited because it's a private repo).

```
Multiple main classes found. Which would you like to run?

Request Actions:
1. com.foo.bar.util.RecordDeserializer
2. com.foo.bar.util.ReadRecords
3. com.foo.bar.util.SysClient
4. com.foo.bar.util.PrintToFile
5. com.foo.bar.util.PushUserData
6. com.foo.bar.util.Playground
Type number and <Enter> or click with the mouse (q or empty cancels):
```

So lets say the second action `ReadRecords` requires two arguments of `YYYY-MM-DD` format to specify which date ranges to look. I want to be able to pass that in as I pick it here. Having to define the arguments (which frequently change) in my `init.lua` or in some sort of local-to-the-repo `.vimrc.local` that loads in isn't ideal.

Any tips or advice on how to achieve the most, uh, frictionless way to call this job? (This might include defining things differently/better in the `build.sbt` than we currently are.)

In IntelliJ I click on a job dropdown and "Edit Configuration" and put the arguments in there before clicking the debug button. I was hoping for a similar interaction prompt here.

I realize this may not be possible right now! Maybe the answer is "Sorry dude. Update the args in the `init.lua` every time." And that's fine. 😃 But if there's a way to more on-the-flly or interactively pass args in (like you can via `sbt doThing arg1 arg2`) I'd prefer that.

**Reply by ckipp01** [reply]
So it really has nothing to do with how your sbt is structured, so don't go down that route to fix this issue.

Honestly, I don't have a great solution for you off the top of my head. I'd have to take some time and try to dig more into `nvim-dap` to see if there is an easy way to do this that I don't know about. To at least point you in the right direction, you can sort of see how this is set up in `nvim-metals` here: https://github.com/scalameta/nvim-metals/blob/5bf06590e9a412bf1c27c8dfacc245733e0929b3/lua/metals/setup.lua#L355-L398

Notice that it's set up by a function that takes a callback and a config. That config is what will actually hold those args under the `metals` key. So if that could be intercepted somehow and be dynamically set, that'd be the way to do it.

**Reply by jsatk** [reply]
This is good info. Sounds like its simply not possible to do when you call it and has to be set via configs before hand.

If I get some time and energy I will look into ways to pass args dynamically like how you do when you call `sbt foo arg1 arg2`. 

Thanks!

**Reply by mfussenegger** [reply]
You can run dynamic configurations directly using `dap.run`. (See `:help dap.run()`)

You could create a custom command or something that let's you define the arguments, creates the `config` definition and calls `dap.run`

**Reply by jsatk** [reply]
> You can run dynamic configurations directly using `dap.run`. (See `:help dap.run()`)
> 
> You could create a custom command or something that let's you define the arguments, creates the `config` definition and calls `dap.run`

Oooo. I hadn't read through the `dap.api` part of the help yet. Thanks. Good ideas. I'm sure it's possible via this some how.

**Reply by mfussenegger** [reply]
Another option would be to use a function as `args` property that returns the arguments.  Inside this function you could prompt for values using `input()`.

Then when you use continue and select that configuration entry it would prompt you for the argument values
